### PR TITLE
[PEAC-880] Feature/register

### PR DESCRIPTION
**Pull request by tuannguyensn2001**
Jira Ticket:
- https://employmenthero.atlassian.net/browse/PEAC-880

Summary:
- Register user in system

Actions:
- Write API for register user
- Writer unit test to verify 

Testing plan
- Setup environment
- Start server
- Send request to /api/v1/auth/register with username, password, email
- Expect new user registered if email haven't existed in system
- Expect error if user existed

### CODE REVIEW

File: spec/controllers/auth_spec.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+          email: "tuannguyensn2001a@gmail.com",
+          username: "tuannguyensn2001",
+          password: "tuannguyensn2001"
+        }
+      end
+      it "return response success" do
+        post :register, params: params
+        expect(response).to have_http_status(:ok)
+        expect(User.count).to eq(1)
+        expect(User.first.email).to eq("tuannguyensn2001a@gmail.com")
```
Comment (hieubui149) [inline spec/controllers/auth_spec.rb:17]: This controller is being tested with the Auth::Register logic when it shouldn't. You have to stub the service

File: spec/factories/user.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
-#   end
-# end
\ No newline at end of file
+FactoryBot.define do
+  factory :user do
+    email { Faker::Internet.email }
+    username { Faker::Internet.username(specifier: 6..8) }
+    password { Faker::Internet.password(min_length: 10) }
+  end
+end
```
Comment (hieubui149) [inline spec/factories/user.rb:7]: Every files must have a newline at the end of file

**Review by hieubui149 — COMMENTED** [review]
You have to practice write PR's description and title to let reviewers know what you intended to do in the PR

File: spec/controllers/auth_spec.rb (hunk 1 above)
Comment (tuannguyensn2001) [reply spec/controllers/auth_spec.rb:17]: you mean that service expect to be called ?

Comment (hieubui149) [reply spec/controllers/auth_spec.rb:17]: yup, stub the service to be called with the correct params and mock the returning data

File: spec/controllers/auth_spec.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    context "with invalid params" do
+      let (:params) do
+        {
+          email: ""
+        }
+      end
+      it "return response error" do
+        post :register, params: params
+        expect(response).not_to have_http_status(:ok)
```
Comment (huyvohcmc) [inline spec/controllers/auth_spec.rb]: Why don't we expect status code to be `bad_request`?

File: spec/controllers/auth_spec.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      end
+    end
+
+    context "with invalid params" do
+      let (:params) do
+        {
+          email: ""
+        }
+      end
+      it "return response error" do
```
Comment (huyvohcmc) [inline spec/controllers/auth_spec.rb]: The test says "return response error", but I don't see we expect the response to have error...

File: spec/services/auth/register_spec.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    context "with invalid params" do
+      let(:invalid_params) { { email: "test@example.com" } }
+
+      it "returns an error" do
+        service = described_class.new(invalid_params)
+        service.call
+
+        # expect(service.errors).to eq(["data not valid"])
+        expect(service.error?).to be_truthy
```
Comment (huyvohcmc) [inline spec/services/auth/register_spec.rb:36]: It's better to explicitly state what error are we expecting in this test.

**Comment by huyvohcmc** [comment]
@tuannguyensn2001 There are a few blank lines and also there are a few "no newline at the end of file", could you address those?

File: spec/controllers/auth_spec.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    context "with invalid params" do
+      let (:params) do
+        {
+          email: ""
+        }
+      end
+      it "return response error with invalid params" do
+        mock_service = instance_double(Auth::Register, call: nil, errors: ["data not valid"])
+        allow(Auth::Register).to receive(:new).and_return(mock_service)
```
Comment (hieubui149) [inline spec/controllers/auth_spec.rb:29]:
```suggestion
        expect(Auth::Register).to receive(:new).and_return(mock_service)
```
also, you missed mock/stub for the happy case above

File: app/controllers/auth_controller.rb [hunk 7]
```diff
@@ -1,15 +1,15 @@
...
-#
-#   end
-# end
+class AuthController < ApplicationController
+  def register
+    request = params.permit(:email, :password, :username)
+
+    service = Auth::Register.new(request)
+    service.call
+    if service.errors.any?
```
Comment (huyvohcmc) [inline app/controllers/auth_controller.rb:7]: Can be simply written as `if service.error?`

**Pull request merged**